yt/frontends/parthenon/data_structures.py

```diff
@@ -0,0 +1,326 @@
+
+    def _populate_grid_objects(self):
+        for g in self.grids:
+            g._prepare_grid()
+            g._setup_dx()
+        self.max_level = self._handle["Info"].attrs["MaxLevel"]
+
+
+class ParthenonDataset(Dataset):
+    _field_info_class = ParthenonFieldInfo
+    _dataset_type = "parthenon"
+
+    def __init__(
+        self,
+        filename,
+        dataset_type="parthenon",
+        storage_filename=None,
+        parameters=None,
+        units_override=None,
+        unit_system="cgs",
+        default_species_fields=None,
+        magnetic_normalization="gaussian",
+    ):
+        self.fluid_types += ("parthenon",)
+        if parameters is None:
+            parameters = {}
+        self.specified_parameters = parameters
+        if units_override is None:
+            units_override = {}
+        self._handle = HDF5FileHandler(filename)
+        xrat = self._handle["Info"].attrs["RootGridDomain"][2]
+        yrat = self._handle["Info"].attrs["RootGridDomain"][5]
+        zrat = self._handle["Info"].attrs["RootGridDomain"][8]
+        if xrat != 1.0 or yrat != 1.0 or zrat != 1.0:
+            self.logarithmic = True
+            raise ValueError(
+                "Logarithmic grids not yet supported in Parthenon frontend."
+            )
+        else:
+            self._index_class = ParthenonHierarchy
+            self.logarithmic = False
+        self._magnetic_factor = get_magnetic_normalization(magnetic_normalization)
+
+        self.geometry = geom_map[self._handle["Info"].attrs["Coordinates"]]
+
+        if self.geometry == "cylindrical":
+            axis_order = ("r", "theta", "z")
+        else:
+            axis_order = None
+
+        Dataset.__init__(
+            self,
+            filename,
+            dataset_type,
+            units_override=units_override,
+            unit_system=unit_system,
+            default_species_fields=default_species_fields,
+            axis_order=axis_order,
+        )
+        if storage_filename is None:
+            storage_filename = self.basename + ".yt"
+        self.storage_filename = storage_filename
+
+    def _set_code_unit_attributes(self):
+        """
+        Generates the conversion to various physical _units based on the
+        parameter file
+        """
+        for unit, cgs in [
+            ("length", "cm"),
+            ("time", "s"),
+            ("mass", "g"),
+        ]:
+            unit_param = f"Hydro/code_{unit}_cgs"
+            # We set these to cgs for now, but they may have been overridden
+            if getattr(self, unit + "_unit", None) is not None:
+                continue
+            elif unit_param in self.parameters:
+                setattr(
+                    self, f"{unit}_unit", self.quan(self.parameters[unit_param], cgs)
+                )
+            else:
+                mylog.warning(f"Assuming 1.0 = 1.0 {cgs}")
+                setattr(self, f"{unit}_unit", self.quan(1.0, cgs))
```

Comment:
This looks a lot like it could be simplified with `yt.funcs.setdefaultattr` (we'd loose the log warning, but I don't think that's essential)

Reply:
If I understand the logic correctly, we can get rid of the first `if` but still need to differentiate between setting the explicit unit from the output file and implicit cgs (and are still able to show a warning), don't we? This is also what I did now.
Otherwise, I'm missing sth.